src/conv/invokers/impl_gemm_dynamic.cpp

```diff
@@ -215,6 +217,184 @@ InvokerFactory MakeImplGemmDynamicBackwardDataInvokerFactory(const ConvolutionCo
+    int dtile_dx      = dilation_w / gcd_stride_dilation_w;
+    int dtile_y       = y_tilda;
+    int dtile_x       = x_tilda;
+    int dtile_h       = h_tilda;
+    int dtile_w       = w_tilda;
+    int dslice_h      = h_tilda_slice;
+    int dslice_w      = w_tilda_slice;
+    int dslice_h_left = h_tilda_left;
+    int dslice_w_left = w_tilda_left;
+    int pack_align    = 0;
+    std::vector<int> dtile_iy_gid;
+    std::vector<int> dtile_ix_gid;
+    std::vector<int> y_dot_slice_gid;
+    std::vector<int> x_dot_slice_gid;
+    std::vector<bool> is_gemm_not_empty;
+    for(int gemm_id = 0; gemm_id < num_of_gemms; gemm_id++)
+    {
+        dtile_iy_gid.emplace_back(gemm_id / x_tilda);
+        dtile_ix_gid.emplace_back(gemm_id % x_tilda);
+        y_dot_slice_gid.emplace_back((dtile_iy_gid[gemm_id] + 1) * y_dot <= y ? y_dot : y % y_dot);
+        x_dot_slice_gid.emplace_back((dtile_ix_gid[gemm_id] + 1) * x_dot <= x ? x_dot : x % x_dot);
+        const int gemm_k_gid = k * y_dot_slice_gid[gemm_id] * x_dot_slice_gid[gemm_id];
+        is_gemm_not_empty.emplace_back(gemm_k_gid > 0);
+    }
+    bool need_set_zero = false;
+    if(y < stride_h || x < stride_w || dilation_h != 1 || dilation_w != 1)
+        need_set_zero = true;
+
+    int nxb = cfg.nxb;
+    int b   = h_tilda_slice * w_tilda_slice;
+    b       = (cfg.nxe == 0) ? (b) : ((b + nxb - 1) / nxb) * nxb; // pad to nxb modulo when nxe != 0
+
+    uint32_t nb_n0          = cfg.tensor_b_cluster_lengths[2] * cfg.tensor_b_thread_lengths[2];
+    uint32_t nb_n1b         = cfg.tensor_b_cluster_lengths[3] * cfg.tensor_b_thread_lengths[3];
+    uint32_t unmerge_sub_n  = cfg.gemm_n_per_block / cfg.nxb;
+    uint32_t unmerge_sub_n1 = unmerge_sub_n / nb_n0;
+
+    magic_div_u32_t mdiv_2 =
+        magic_div_u32_gen(((c / group) * n * b) / (cfg.gemm_m_per_block * cfg.gemm_n_per_block));
+    magic_div_u32_t mdiv_3 = magic_div_u32_gen((n * b) / cfg.gemm_n_per_block);
+    magic_div_u32_t mdiv_4 = magic_div_u32_gen(b * unmerge_sub_n1 / nb_n1b);
+    magic_div_u32_t mdiv_5 = magic_div_u32_gen(b);
+    magic_div_u32_t mdiv_6 = magic_div_u32_gen(w_tilda_slice);
+
+    return [=](const std::vector<Kernel>& kernels) {
+        const auto kernel = kernels[0];
+        return [=](const Handle& handle, const AnyInvokeParams& primitive_parameters) {
+            decltype(auto) data_ctx = primitive_parameters.CastTo<conv::DataInvokeParams>();
+            const auto& tensors     = data_ctx.tensors;
+            float elapsed           = 0;
+            if(need_set_zero)
+            {
+                float zero = 0.f;
+                SetTensor(handle, tensors.outDesc, tensors.out, &zero);
+
+                if(handle.IsProfilingEnabled())
+                    elapsed += handle.GetKernelTime();
+            }
+            for(int gemm_id = 0; gemm_id < num_of_gemms; gemm_id++)
+            {
+                if(is_gemm_not_empty[gemm_id])
+                {
+                    magic_div_u32_t mdiv_0 =
```

Comment:
Can mdiv_0 be computed out of the invoker? `magic_div_u32_gen` computes a integer division.

Reply:
It has been moved.